go/felix/intdataplane/endpoint_mgr_test.go

```diff
@@ -0,0 +1,375 @@
+		ipv6 = "2001:db8::10.0.240.10"
+	)
+	var (
+		epMgr          *endpointManager
+		filterTable    *mockTable
+		rrConfigNormal rules.Config
+		ipVersion      int
+		eth0Addrs      set.Set
+		loAddrs        set.Set
+	)
+
+	BeforeEach(func() {
+		rrConfigNormal = rules.Config{
+			IPIPEnabled:          true,
+			IPIPTunnelAddress:    nil,
+			IPSetConfigV4:        ipsets.NewIPVersionConfig(ipsets.IPFamilyV4, "cali", nil, nil),
+			IPSetConfigV6:        ipsets.NewIPVersionConfig(ipsets.IPFamilyV6, "cali", nil, nil),
+			IptablesMarkAccept:   0x8,
+			IptablesMarkNextTier: 0x10,
+		}
+		eth0Addrs = set.New()
+		eth0Addrs.Add(ipv4)
+		eth0Addrs.Add(ipv6)
+		loAddrs = set.New()
+		loAddrs.Add("127.0.1.1")
+		loAddrs.Add("::1")
+	})
+
+	JustBeforeEach(func() {
+		renderer := rules.NewRenderer(rrConfigNormal)
+		filterTable = newMockTable()
+		epMgr = newEndpointManager(
+			filterTable,
+			renderer,
+			nil,
+			uint8(ipVersion),
+			[]string{"cali"},
+			nil,
+		)
+	})
+
+	for ipVersion = range []uint8{4, 6} {
+		It("should be constructable", func() {
+			Expect(epMgr).ToNot(BeNil())
+		})
+
+		configureHostEp := func(id string, name string, ipv4Addrs []string, ipv6Addrs []string) func() {
+			return func() {
+				epMgr.OnUpdate(&proto.HostEndpointUpdate{
+					Id: &proto.HostEndpointID{
+						EndpointId: id,
+					},
+					Endpoint: &proto.HostEndpoint{
+						Name:              name,
+						ProfileIds:        []string{},
+						Tiers:             []*proto.TierInfo{},
+						ExpectedIpv4Addrs: ipv4Addrs,
+						ExpectedIpv6Addrs: ipv6Addrs,
+					},
+				})
+				epMgr.CompleteDeferredWork()
+			}
+		}
+
+		expectChainsFor := func(name string) func() {
+			return func() {
+				filterTable.checkChains(append(wlDispatchEmpty, hostDispatchForIface(name)...))
+			}
+		}
+
+		expectEmptyChains := func() func() {
+			return func() {
+				filterTable.checkChains(append(wlDispatchEmpty, hostDispatchEmpty...))
+			}
+		}
+
+		removeHostEp := func(id string) func() {
+			return func() {
+				epMgr.OnUpdate(&proto.HostEndpointRemove{
+					Id: &proto.HostEndpointID{
+						EndpointId: id,
+					},
+				})
+				epMgr.CompleteDeferredWork()
+			}
+		}
+
+		Context("with host interfaces eth0, lo", func() {
+			JustBeforeEach(func() {
+				epMgr.OnUpdate(&ifaceUpdate{
+					Name:  "eth0",
+					State: "up",
+				})
+				epMgr.OnUpdate(&ifaceAddrsUpdate{
+					Name:  "eth0",
+					Addrs: eth0Addrs,
+				})
+				epMgr.OnUpdate(&ifaceUpdate{
+					Name:  "lo",
+					State: "up",
+				})
+				epMgr.OnUpdate(&ifaceAddrsUpdate{
+					Name:  "lo",
+					Addrs: loAddrs,
+				})
+				epMgr.CompleteDeferredWork()
+			})
+
+			It("should have empty dispatch chains", func() {
+				filterTable.checkChains(append(wlDispatchEmpty, hostDispatchEmpty...))
+			})
+
+			Describe("with host endpoint matching eth0", func() {
+				JustBeforeEach(configureHostEp("id1", "eth0", []string{}, []string{}))
+				It("should have expected chains", expectChainsFor("eth0"))
+
```

Comment:
Unless it's covered subtly and I missed it, one case that may be missing here is checking the precedence when you have multiple matches (i.e. checking that the best match is stable)